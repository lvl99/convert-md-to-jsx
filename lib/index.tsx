import React, { createElement, JSX } from "react";
import {
  BlockContentMap,
  Blockquote,
  Break,
  Code,
  Definition,
  Delete,
  Emphasis,
  FootnoteDefinition,
  FootnoteReference,
  Heading,
  Html,
  Image,
  ImageReference,
  InlineCode,
  Link,
  LinkReference,
  List,
  ListItem,
  Node,
  Paragraph,
  Parent,
  PhrasingContentMap,
  Root,
  RootContentMap,
  Strong,
  Table,
  TableCell,
  TableRow,
  Text,
  ThematicBreak,
  Yaml,
} from "mdast";
import { gfm } from "micromark-extension-gfm";
import { gfmFromMarkdown } from "mdast-util-gfm";
import { fromMarkdown } from "mdast-util-from-markdown";

export interface GenericProps {
  [key: string]: unknown;
}

export type JSXComponent = JSX.ElementType;
export type JSXNode = JSX.Element;

export type ASTNodeType =
  | keyof BlockContentMap
  | keyof RootContentMap
  | keyof PhrasingContentMap;
export type ASTNode =
  | Node
  | Parent
  | Blockquote
  | Code
  | Html
  | Heading
  | ThematicBreak
  | Break
  | Definition
  | Delete
  | Emphasis
  | FootnoteReference
  | FootnoteDefinition
  | Image
  | ImageReference
  | InlineCode
  | Link
  | LinkReference
  | Strong
  | ListItem
  | List
  | Paragraph
  | Root
  | Table
  | TableRow
  | TableCell
  | Text
  | Yaml;

export interface ASTNodeProps {
  node: ASTNode;
  parentNode?: ASTNode;
  children?: JSXNode[] | null;
}

export interface ASTNodeMap {
  [key: string]: JSXComponent;
}

export const DEFAULT_NODE_MAP: ASTNodeMap = Object.freeze({
  blockquote: ({ children }) => <blockquote>{children}</blockquote>,
  break: () => <br />,
  code: ({ node }) =>
    ("value" in node && (
      <pre>
        <code>{node.value}</code>
      </pre>
    )) ||
    null,
  definition: () => null,
  delete: ({ children }) => <s>{children}</s>,
  emphasis: ({ children }) => <em>{children}</em>,
  footnoteReference: ({ node }) =>
    ("identifier" in node && (
      <sup>
        <a href={"#fn-def-" + node.identifier}>
          {("label" in node && node.label) || node.identifier}
        </a>
      </sup>
    )) ||
    null,
  footnoteDefinition: ({ node, children }) =>
    "identifier" in node && (
      <p id={"fn-def-" + node.identifier}>
        {("label" in node && node.label) || node.identifier}: {children}
      </p>
    ),
  heading: ({ node, children }) => {
    const depth = ("depth" in node && node.depth) || 1;
    switch (depth) {
      case 1:
        return <h1>{children}</h1>;
      case 2:
        return <h2>{children}</h2>;
      case 3:
        return <h3>{children}</h3>;
      case 4:
        return <h4>{children}</h4>;
      case 5:
        return <h5>{children}</h5>;
      case 6:
        return <h6>{children}</h6>;
      default:
        return <p>{children}</p>;
    }
  },
  html: ({ node }) => ("value" in node && node.value) || null,
  image: ({ node }) =>
    ("url" in node && (
      <img src={node.url} title={"title" in node ? node.title || "" : ""} />
    )) ||
    null,
  imageReference: () => null,
  inlineCode: ({ node }) =>
    ("value" in node && <code>{node.value}</code>) || null,
  link: ({ node, children }) =>
    "url" in node ? <a href={node.url}>{children}</a> : children,
  linkReference: () => null,
  list: ({ node, children }) => {
    const ordered = "ordered" in node && node.ordered;
    if (ordered) {
      return <ol>{children}</ol>;
    } else {
      return <ul>{children}</ul>;
    }
  },
  listItem: ({ node, parentNode, children }) => (
    <li>
      {"checked" in node && typeof node.checked === "boolean"
        ? (node.checked && (
            <input type="checkbox" checked={node.checked} disabled={true} />
          )) || <input type="checkbox" disabled={true} />
        : null}
      {children}
    </li>
  ),
  paragraph: ({ parentNode, children }) =>
    parentNode?.type === "listItem" ||
    parentNode?.type === "footnoteDefinition" ? (
      children
    ) : (
      <p>{children}</p>
    ),
  root: ({ children }) => <>{children}</>,
  strong: ({ children }) => <strong>{children}</strong>,
  table: ({ children }) => <table>{children}</table>,
  tableRow: ({ children }) => <tr>{children}</tr>,
  tableCell: ({ children }) => <td>{children}</td>,
  text: ({ node }) => ("value" in node && node.value) || null,
  thematicBreak: () => <hr />,
  yaml: ({ node }) =>
    ("value" in node && (
      <pre>
        <code>{node.value}</code>
      </pre>
    )) ||
    null,
});

/**
 * Convert AST Node to React Component
 */
export const convertNodeToJSX = (
  node: ASTNode,
  nodeMap: ASTNodeMap = DEFAULT_NODE_MAP,
  parentNode?: ASTNode,
): JSXNode => {
  if (node.type in nodeMap) {
    const element = nodeMap[node.type];
    if ("children" in node) {
      const children = Array.from(node.children).map((childNode: ASTNode) =>
        convertNodeToJSX(childNode, nodeMap, node),
      );
      return createElement(element, { node, parentNode }, ...children);
    } else {
      return createElement(element, { node, parentNode });
    }
  } else {
    throw new Error(`ASTNode transformer not mapped for type: ${node.type}`);
  }
};

/**
 * Convert MD node to JSX (ReactNode).
 */
export const convertMDToJSX = (
  node: string,
  nodeMap: ASTNodeMap = DEFAULT_NODE_MAP,
) => {
  const root = fromMarkdown(node, {
    extensions: [gfm()],
    mdastExtensions: [gfmFromMarkdown()],
  });
  const output = convertNodeToJSX(root, nodeMap);
  return output;
};
