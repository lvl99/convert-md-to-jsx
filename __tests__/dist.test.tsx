import React, { memo } from "react";
import { renderToStaticMarkup } from "react-dom/server";
import { describe, it, expect } from "vitest";
import convert, { DEFAULT_NODE_MAP } from "../dist";
import { JSXComponent, ASTNodeProps } from "../dist/lib";

describe("convert", () => {
  it("should work as advertised", () => {
    const test = `# Heading

This is a [test](https://www.google.com).`;
    const result = convert(test);
    expect(result).toHaveProperty("type", DEFAULT_NODE_MAP.root);
    expect(result).toHaveProperty("props.children");
    // I'm a loser baby, so why don't ya kill meeee~~~
    // @ts-ignore
    expect(result.props.children).not.toBeNull();
  });

  it("should work as advertised with custom nodeMap", () => {
    const test = `# Heading

This is a [test](https://www.google.com).`;
    // Ugh, this sux
    const Root: JSXComponent<ASTNodeProps> = ({ children }) => (
      <div className="hello">{children}</div>
    );
    const result = convert(test, {
      ...DEFAULT_NODE_MAP,
      root: Root,
    });
    expect(result).toHaveProperty("type", Root);
    expect(result).toHaveProperty("props.children");
    // I'm a loser baby, so why don't ya kill meeee~~~
    // @ts-ignore
    expect(result.props.children).not.toBeNull();
  });
});

describe("react", () => {
  it("should render MD to JSX to HTML", () => {
    const test = `# Heading 1

This is a [test](https://www.google.com).

Some *text* **formatting** ***here***. ~Ignore this!~ =Look at this!=

---

## Heading 2

### Heading 3

#### Heading 4

##### Heading 5

###### Heading 6

- Item 1
- Item 2
- Item 3

1. Item 1
2. Item 2
3. Item 3

> Blockquote

HTML code: <img src="https://placehold.it/600x600">

![image](https://placehold.it/600x600)

This is \`inline code\`.

\`\`\`
This is a code block.
\`\`\`

# GFM

- [ ] Unchecked item
- [x] Checked item

| Column | Column |
| ------ | ------ |
| Cell   | Cell   |

Here's my footnote in use[^1].

Term
: Definition

[^1]: This is the footnote definition.
`;

    const RenderMarkdown = memo(({ input }: { input: string }) =>
      convert(input),
    );

    const result = renderToStaticMarkup(<RenderMarkdown input={test} />);
    expect(result).toMatchSnapshot();
  });
});
