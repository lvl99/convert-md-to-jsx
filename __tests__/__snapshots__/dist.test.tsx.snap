// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react > should render MD to JSX to HTML 1`] = `
"<h1>Heading 1</h1><p>This is a <a href="https://www.google.com">test</a>.</p><p>Some <em>text</em> <strong>formatting</strong> <em><strong>here</strong></em>. <s>Ignore this!</s> =Look at this!=</p><hr/><h2>Heading 2</h2><h3>Heading 3</h3><h4>Heading 4</h4><h5>Heading 5</h5><h6>Heading 6</h6><ul><li>Item 1</li><li>Item 2</li><li>Item 3</li></ul><ol><li>Item 1</li><li>Item 2</li><li>Item 3</li></ol><blockquote><p>Blockquote</p></blockquote><p>HTML code: &lt;img src=&quot;https://placehold.it/600x600&quot;&gt;</p><p><img src="https://placehold.it/600x600" title=""/></p><p>This is <code>inline code</code>.</p><pre><code>This is a code block.</code></pre><h1>GFM</h1><ul><li><input type="checkbox" disabled=""/>Unchecked item</li><li><input type="checkbox" disabled="" checked=""/>Checked item</li></ul><table><tr><td>Column</td><td>Column</td></tr><tr><td>Cell</td><td>Cell</td></tr></table><p>Here&#x27;s my footnote in use<sup><a href="#fn-def-1">1</a></sup>.</p><p>Term
: Definition</p><p id="fn-def-1">1: This is the footnote definition.</p>"
`;
